components/ws-manager/pkg/manager/internal/grpcpool/pool.go

```diff
@@ -60,79 +49,85 @@ func (p *Pool) Get(host string) (*grpc.ClientConn, error) {
 		p.mu.RUnlock()
 		return nil, ErrPoolClosed
 	}
+
 	conn, exists := p.connections[host]
 	p.mu.RUnlock()
 
-	if !exists || conn.GetState() == connectivity.Shutdown {
-		return p.add(host)
+	if !exists {
+		return nil, ErrNoWSDeamonFoundInPool
 	}
 
-	return conn, nil
+	c, err := conn.Get()
+	if err != nil {
+		return nil, xerrors.Errorf("cannot create connection to ws-daemon: %w", err)
+	}
+
+	return c.Value(), nil
 }
 
 // add adds a new connection to the host if one doesn't exist already in a state that is not Shutdown.
 // Compared to Get, this function holds a write lock on mu. Get uses this function if it cannot find
 // an existing connection.
-func (p *Pool) add(host string) (*grpc.ClientConn, error) {
+func (p *Pool) Add(host, address string) error {
 	p.mu.Lock()
 	defer p.mu.Unlock()
 
-	conn, exists := p.connections[host]
-	if exists && conn.GetState() != connectivity.Shutdown {
-		return conn, nil
+	_, exists := p.connections[host]
+	if exists {
+		return nil
 	}
 
-	conn, err := p.factory(host)
+	connPool, err := conn_pool.New(address, conn_pool.Options{
+		Dial:                 p.dial,
+		MaxIdle:              8,
+		MaxActive:            64,
+		MaxConcurrentStreams: 64,
+		Reuse:                false,
+	})
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	p.connections[host] = conn
-	return conn, nil
+	p.connections[host] = connPool
+
+	_, err = connPool.Get()
+	if err != nil {
+		return err
+	}
+
+	return nil
 }
 
-// Close empties the pool after closing all connections it held.
-// It waits for all connections to close.
-// Once the pool is closed, calling Get will result in ErrPoolClosed
-func (p *Pool) Close() error {
+func (p *Pool) Remove(host string) error {
 	p.mu.Lock()
 	defer p.mu.Unlock()
 
-	p.closed = true
-	errs := make([]string, 0)
-	for _, c := range p.connections {
-		err := c.Close()
-		if err != nil {
-			errs = append(errs, err.Error())
-		}
+	connPool, exists := p.connections[host]
+	if !exists {
+		return nil
 	}
 
-	if len(errs) != 0 {
-		return fmt.Errorf("pool close: %s", strings.Join(errs, "; "))
-	}
+	delete(p.connections, host)
 
-	return nil
+	return connPool.Close()
 }
 
-// ValidateConnections check if existing connections in the pool
-// are using valid addresses and remove them from the pool if not.
-func (p *Pool) ValidateConnections() {
-	p.mu.RLock()
-	addresses := make([]string, 0, len(p.connections))
-	for address := range p.connections {
-		addresses = append(addresses, address)
-	}
-	p.mu.RUnlock()
+// Close empties the pool after closing all connections it held.
+// It waits for all connections to close.
+// Once the pool is closed, calling Get will result in ErrPoolClosed
+func (p *Pool) Stop() {
+	p.mu.Lock()
+	defer p.mu.Unlock()
 
-	for _, address := range addresses {
-		if p.isValidConnection(address) {
-			continue
-		}
+	if p.closed {
+		p.mu.RUnlock()
```

Comment:
Why do we need to unlock it again? We have `defer p.mu.Unlock()` above.

Reply:
fixed

Reply:
🤔 The code still executes `p.mu.RUnlock()`.

/hold